exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go

```diff
@@ -0,0 +1,308 @@
+
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
+}
+
+var _ AggregatableResponse = &BeforeClusterCreateResponse{}
+
+// BeforeClusterCreateResponse is the response of BeforeClusterCreate hook.
+// +kubebuilder:object:root=true
+type BeforeClusterCreateResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Status of the call. One of "Success" or "Failure".
+	Status ResponseStatus `json:"status"`
+
+	// RetryAfterSeconds when set to a non-zero signifies that the hook
+	// will be called again at a future time.
+	RetryAfterSeconds int `json:"retryAfterSeconds"`
+
+	// A human-readable description of the status of the call.
+	Message string `json:"message"`
+}
+
+// Aggregate combines individual extension handler responses into a single response of the hook.
+func (r *BeforeClusterCreateResponse) Aggregate(responses []*ExtensionHandlerResponse) error {
+	if err := aggregateExtensionResponses(r, responses); err != nil {
+		return errors.Wrap(err, "failed to aggregate extension responses")
+	}
+	return nil
+}
+
+// BeforeClusterCreate is the runtime hook that will be called right before a Cluster is created.
+func BeforeClusterCreate(*BeforeClusterCreateRequest, *BeforeClusterCreateResponse) {}
+
+// AfterControlPlaneInitializedRequest is the request of the hook.
+// +kubebuilder:object:root=true
+type AfterControlPlaneInitializedRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
+}
+
+var _ AggregatableResponse = &AfterControlPlaneInitializedResponse{}
+
+// AfterControlPlaneInitializedResponse is the response of AfterControlPlaneInitialized hook.
+// +kubebuilder:object:root=true
+type AfterControlPlaneInitializedResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Status of the call. One of "Success" or "Failure".
+	Status ResponseStatus `json:"status"`
+
+	// A human-readable description of the status of the call.
+	Message string `json:"message"`
+}
+
+// Aggregate combines individual extension handler responses into a single response of the hook.
+func (r *AfterControlPlaneInitializedResponse) Aggregate(responses []*ExtensionHandlerResponse) error {
+	if err := aggregateExtensionResponses(r, responses); err != nil {
+		return errors.Wrap(err, "failed to aggregate extension responses")
+	}
+	return nil
+}
+
+// AfterControlPlaneInitialized is the runtime hook that will be called after the first control plane is available.
+func AfterControlPlaneInitialized(*AfterControlPlaneInitializedRequest, *AfterControlPlaneInitializedResponse) {
+}
+
+// BeforeClusterUpgradeRequest is the request of the hook.
+// +kubebuilder:object:root=true
+type BeforeClusterUpgradeRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
+
+	// The current version of the cluster.
```

Comment:
```suggestion
	// The current Kubernetes version of the cluster.
```
nit (same in other places)